### ISSUE TITLE
Inbox listener doesn't know to which inbox it delivers

### ISSUE THREAD

**Issue by mattes3** [dahlia/fedify#117]
I am working on a listener for `Create` activities. The problem is that I do not know where to deliver a Note because the addressing looks like this:

```
  to: URL 'https://www.w3.org/ns/activitystreams#Public',
  cc: URL 'https://example.com/ap/members/alfred/followers',
```

If I knew which inbox triggered the `.on(Create...)` callback, I would simply deliver it to the user who owns that inbox.

I cannot enumerate the followers inside the inbox listener because this way I would deliver the Note several times.

What should I do to find out the target user?

**Comment by dahlia** [comment]
Due to [shared inbox](https://www.w3.org/TR/activitypub/#shared-inbox-delivery), you need to rely on the `to`/`cc` properties to get the target of the `Create` activity. If you're worried about delivering the same activity multiple times, you can specify the same `id` for the same activity.

**Comment by mattes3** [comment]
🤔 this is weird. When the sending server uses the named inboxes, I possibly get 100 activities, one for each follower’s inbox. Then I enumerate the 100 followers again and distribute the same Note object 10,000 times.

This cannot be the idea, right?

**Comment by dahlia** [comment]
Yeah, that's why we have a shared inbox.  By the way, I wonder what's your specific scenario.  I could help you.

**Comment by mattes3** [comment]
My specific scenario is: I am writing an app that behaves like a combination of Lemmy and Mastodon: You have forums with threads, and you have users. Users can follow forums or single users (on Lemmy you can only follow a forum, not a single user).

Now, in this case, I sent a Create activity using Fedify:

```typescript
await ctx.sendActivity({ handle: senderHandle }, 'followers', createActivity, {
    preferSharedInbox: true,
});
```

Fedify used the named inboxes (although I told it to use the shared inbox!),  that's why I thought I would get away with this simple addressing:

```
to: PUBLIC_COLLECTION,
cc: ctx.getFollowersUri(senderHandle),
```

I can now see that this will not work because the receiving server (in this case my server as well, but could be any server) does not know the sender's follower collection (the sender is on my server).

Does this mean I will have to put a potentially large collection of follower addresses into the `'cc'` field?

**Comment by dahlia** [comment]
> Fedify used the named inboxes (although I told it to use the shared inbox!),

That's probably because your followers dispatcher do not return enough information (lacking `endpoints.sharedInbox` property), or the recipients actually don't have their shared inboxes.

> I can now see that this will not work because the receiving server (in this case my server as well, but could be any server) does not know the sender's follower collection (the sender is on my server).

You would be able to get the sender's follower collection via [`await activity.getCc()`](https://jsr.io/@fedify/fedify@0.13.1/doc/~/Object.prototype.getCc) (which will return `Collection`).

However, the receiving server should store the following–follower pairs for efficiency; the server could store a pair when it receives a `Follow` activity.  That's how Mastodon and etc work.

**Comment by mattes3** [comment]
Two times "aha!", thank you. 😄 

1) My followers dispatcher returned Recipient objects that didn't have the `endpoints.sharedInbox` property. Fixed. Consequently, Fedify posts to the shared inbox now, that's good.

2) My server is already storing the following–follower pairs for efficiency, so I have a subset of the sender's followers collection, i.e. those followers that reside on my own server. This would be sufficient to solve the problem.

Thanks a lot, I will implement this, and it should work! 👏

**Comment by dahlia** [comment]
Glad to hear it works out for you!
